src/audio/codec_adapter/codec/waves.c

```diff
@@ -0,0 +1,681 @@
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	waves_codec_data_clear(waves_codec);
+	codec->private = waves_codec;
+
+	status = MaxxEffect_GetEffectSize(&waves_codec->effect_size);
+	if (status) {
+		comp_err(dev, "waves_codec_init() MaxxEffect_GetEffectSize() error %d",
+			status);
+		ret = status;
+		codec_free_memory(dev, waves_codec);
+		goto err;
+	}
+
+	waves_codec->effect = (MaxxEffect_t *)codec_allocate_memory(dev,
+		waves_codec->effect_size, 16);
+
+	if (!waves_codec->effect) {
+		comp_err(dev, "waves_codec_init() allocate %d bytes for effect",
+			waves_codec->effect_size);
+		codec_free_memory(dev, waves_codec);
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	comp_info(dev, "waves_codec_init() allocated %d bytes for effect",
+		waves_codec->effect_size);
+
+	comp_dbg(dev, "waves_codec_init() done");
+	return 0;
+
+err:
+	comp_err(dev, "waves_codec_init() failed %d", ret);
+	return ret;
+}
+
+int waves_codec_prepare(struct comp_dev *dev)
+{
+	int ret = 0;
+
+	comp_dbg(dev, "waves_codec_prepare() start");
+
+	ret = waves_effect_check(dev);
+	if (ret)
+		goto err;
+
+	ret = waves_effect_init(dev);
+	if (ret)
+		goto err;
+
+	ret = waves_effect_allocate(dev);
+	if (ret)
+		goto err;
+
+	ret = waves_effect_setup_config(dev);
+	if (ret)
+		goto err;
+
+	comp_dbg(dev, "waves_codec_prepare() done");
+	return 0;
+
+err:
+	comp_err(dev, "waves_codec_prepare() failed %d", ret);
+	return ret;
+}
+
+int waves_codec_process(struct comp_dev *dev)
+{
+	int ret;
+	struct codec_data *codec = comp_get_codec(dev);
+	struct waves_codec_data *waves_codec = codec->private;
+
+	comp_dbg(dev, "waves_codec_process() start");
+
+	MaxxStream_t *i_streams[1] = { &waves_codec->i_stream };
+	MaxxStream_t *o_streams[1] = { &waves_codec->o_stream };
+	MaxxStatus_t status;
+	uint32_t num_input_samples = waves_codec->buffer_samples;
+
+
+	// here input buffer should always be filled up as requested
+	// since noone updates it`s size except code in prepare
+	// kinda odd, but this is how codec adapter operates
+	// on the other hand there is available/produced counters in cpd, so we check them anyways
+	if (codec->cpd.avail != waves_codec->buffer_bytes) {
+		comp_warn(dev, "waves_codec_process() input buffer %d is not full %d",
+			codec->cpd.avail, waves_codec->buffer_bytes);
+		num_input_samples = codec->cpd.avail /
+			(waves_codec->sample_size_in_bytes * waves_codec->i_format.numChannels);
+	}
+
+	waves_codec->i_stream.buffersArray = &waves_codec->i_buffer;
+	waves_codec->i_stream.numAvailableSamples = num_input_samples;
+	waves_codec->i_stream.numProcessedSamples = 0;
+	waves_codec->i_stream.maxNumSamples = waves_codec->buffer_samples;
+
+	waves_codec->o_stream.buffersArray = &waves_codec->o_buffer;
+	waves_codec->o_stream.numAvailableSamples = 0;
+	waves_codec->o_stream.numProcessedSamples = 0;
+	waves_codec->o_stream.maxNumSamples = waves_codec->buffer_samples;
+
+	status = MaxxEffect_Process(waves_codec->effect, i_streams, o_streams);
+	if (status) {
+		ret = status;
```

Comment:
will Waves errors translate into -ERR codes?

Reply:
actually no, thanks for pointing it out
changed so all internal error codes are logged and -ERR codes are propagated